File: pset6/credit.py
from cs50 import get_int
from math import floor

def main():
    cc_num = get_int("Number: ")
    dig1 = 0
    dig2 = 0
    digCounter = 0
    evens = 0
    odds = 0

    while cc_num > 0:
        dig2 = dig1
        dig1 = cc_num % 10

        if digCounter % 2 == 0:
            evens += dig1
        else:
            mult = dig1*2
            odds += (mult // 10) + (mult % 10)

        cc_num //= 10
        digCounter += 1

    is_valid = (evens + odds) % 10 == 0
    first_two_digits = (dig1 * 10) + dig2

    if dig1 == 4 and digCounter >= 13 and digCounter <= 16 and is_valid:
        print("VISA")

    elif first_two_digits >= 51 and first_two_digits <= 55 and digCounter == 16 and is_valid:
        print("MASTERCARD")

    elif (first_two_digits == 34 or first_two_digits == 37) and digCounter == 15 and is_valid:
        print("AMEX")

    else:
        print("INVALID")

if __name__ == "__main__":
    main()

File: pset1/credit.c
//Inputs credit card number, outputs type of card (Visa, Amex, Mastercard)

#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long cc_num = get_long("Number: ");
    long multM = 10;
    long sumM = 0;
    long digM = 0;
    long digLM = 0;
    long digRM = 0;
    long firstDig = 0;
    long secondDig = 0;
    long length = 0;

    do
    {
        digM = (cc_num / multM % 10) * 2;
        if (digM > 9)
        {
            digLM = digM / 10 % 10;
            digRM = digM % 10;
            sumM = sumM + digLM + digRM;
        }
        else
        {
            sumM = sumM + digM;
        }
        if (cc_num / multM < 10 && cc_num / multM != 0)
        {
            firstDig = digM / 2;
        }
        else if (cc_num / multM < 100 && cc_num / multM >= 10)
        {
            secondDig = digM / 2;
        }
        multM = multM * 100;
        length = length + 1;
        //printf("%li\n", digM/2);
    }
    while (cc_num / multM != 0);
    //printf("%li\n", firstDig);
    long mult = 1;
    long sum = 0;
    long dig = 0;

    do
    {
        dig = (cc_num / mult % 10);
        sum = sum + dig;
        if (cc_num / mult < 10 && cc_num / mult != 0)
        {
            firstDig = dig;
        }
        else if (cc_num / mult < 100 && cc_num / mult >= 10)
        {
            secondDig = dig;
        }
         mult = mult * 100;
        length = length + 1;
        //printf("%li\n", dig);
    }
    while (cc_num / mult != 0);
    long totSum = sum + sumM;

    if (length == 15 && firstDig == 3 && totSum % 10 == 0)
    {
        if (secondDig == 4 || secondDig == 7)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (length == 16 && firstDig == 5 && totSum % 10 == 0)
    {
        if (secondDig == 1 || secondDig == 2 || secondDig == 3 || secondDig == 4 || secondDig == 5)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (firstDig == 4 && totSum % 10 == 0)
    {
        if (length == 13 || length == 16)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
